corehq/messaging/util.py

```diff
@@ -41,14 +49,26 @@ def rule_initiation_key_is_set(self):
     def rule_initiation_key_minutes_remaining(self):
         return (self.client.ttl(self.rule_initiation_key) // 60) or 1
 
-    def set_initial_progress(self):
+    def set_initial_progress(self, shard_count=0):
+        # shard_count is passed when tasks run pear each shard
         self.client.set(self.current_key, 0)
         self.client.set(self.total_key, 0)
+        if shard_count:
+            self.client.set(self.shard_count_key, shard_count)
         self.client.expire(self.current_key, self.key_expiry)
         self.client.expire(self.total_key, self.key_expiry)
         self.client.delete(self.rule_cancellation_key)
         self.set_rule_initiation_key()
 
+    def mark_shard_complete(self, db_alias):
+        completed_shards = self.client.get(self.completed_shards_key) or []
+        self.client.set(self.completed_shards_key, completed_shards + ['db_alias'])
```

Comment:
I think this meant to be variable `db_alias`?

Reply:
Also other methods that update the counter, reset the expiry for the key, should this do that too? otherwise it will get removed.

Reply:
I am not sure there is a user of keeping this as a list as oppose to just count of shards completed. Using `incr` is anymore reliable since this can run into a very very edgy case where its updated for two shards in parallel ending up with skipping a shard in this list post completion

Reply:
The main reason to track the shards is so that we have an idea which shards are all finished and which ones remaining in case we need to debug.

Reply:
Regarding expiring, I guess instead of expiring, I will just delete them in set_rule_complete, that should work, yeah?

Reply:
Did this.

Reply:
If i remember correctly the default expiry is 5 minutes, so we will lose that value if we not set and update the expiry on that key, yes?

Reply:
Updated

Reply:
Cool, thanks